BatteryManagementSystem.cpp

```diff
@@ -0,0 +1,93 @@
+/**************************************************************************************
+f* @file        : BatteryManagementSystem.cpp
+* @brief        : File contains necessary business logic to handle core logic of BMS
+**************************************************************************************/
+
+#include <assert.h>
+#include <iostream>
+
+#include "BatteryManagementSystem.h"
+#include "PrintOutput.h"
+
+using namespace std;
+using namespace BatteryManagementSystem;
+
+BMS::BMS()
+{
+}
+
+BMS::~BMS()
+{
+	umap_BMSData.clear();
+}
+
+/**
+ * Description     : AddParameter : Adds the Parameter and its info to BMS map, if parameter not added already
+ *
+ */
+bool BMS::AddParameter(BatteryParameter parameter,ParameterInfo data)
+{
+	if(umap_BMSData.find(parameter) == umap_BMSData.end())
+	{
+		umap_BMSData[parameter] = data;
+		return true;
+	}
+	return false;
+}
+
+/**
+ * Description     : RemoveParameter : Removes the Parameter and its info from BMS map, if parameter present in map
+ *
+ */
+bool BMS::RemoveParameter(BatteryParameter parameter)
+{
+	if(umap_BMSData.find(parameter) != umap_BMSData.end())
+	{
+		umap_BMSData.erase(parameter);
+		return true;
+	}
+	return false;
+}
+
+/**
+ * Description     : checkBatteryParameter_withValue : Tests the Battery Parameter State based on configuration, with the input Value.
+ * 					 Returns false if lesser than min threshold or greater than max threshold. Returns true if within range.
+ *
+ */
+bool BMS::checkBatteryParameter_withValue(BatteryParameter parameter,float currentValue)
+{
+	return isParamWithinRange(parameter, currentValue);
+}
+
+/**
+ * Description     : isParamWithinRange : Tests the Battery Parameter with the input Value.
+ * 					 Prints Output and returns false if lesser than min threshold or greater than max threshold. Returns true if within range or same as threshold
+ *
+ */
+bool BMS::isParamWithinRange(BatteryParameter parameter,float currentValue)
+{
+	BMSMap::iterator itr = umap_BMSData.find(parameter);
+	if(itr != umap_BMSData.end())
+	{
+		Bounds bound = checkBounds(currentValue, itr->second.min, itr->second.max);
+		if(bound == Normal)
+		{
+			return true;
+		}
+		printOutput(itr->second, bound, currentValue);
```

Comment:
The `second` over here is not obvious, and may lead to mistakes when more details are added to BMSData. Is there any way to make that clear?